src/core/server/status/plugins_status.ts

```diff
@@ -57,114 +97,139 @@ export class PluginsStatusService {
         `Custom statuses cannot be registered after setup, plugin [${plugin}] attempted`
       );
     }
-    this.pluginStatuses.set(plugin, status$);
-    this.update$.next(true); // trigger all existing Observables to update from the new source Observable
+
+    const subscription = this.pluginReportedStatus[plugin];
+    if (subscription) subscription.unsubscribe();
+
+    this.pluginReportedStatus[plugin] = status$.subscribe((status) => {
+      // console.log('⭐ Reported status!', plugin, status.level.toString());
+      const previousReportedLevel = this.pluginData[plugin].reportedStatus?.level;
+
+      this.pluginData[plugin].reportedStatus = status;
+      this.pluginStatus[plugin] = status;
+
+      if (status.level !== previousReportedLevel) {
+        this.updatePluginStatuses$.next(plugin);
+      }
+    });
   }
 
   public blockNewRegistrations() {
     this.newRegistrationsAllowed = false;
   }
 
   public getAll$(): Observable<Record<PluginName, ServiceStatus>> {
-    if (!this.allPluginsStatuses) {
-      this.allPluginsStatuses = this.getPluginStatuses$([...this.deps.pluginDependencies.keys()]);
-    }
-    return this.allPluginsStatuses;
+    return this.pluginStatus$.asObservable();
   }
 
   public getDependenciesStatus$(plugin: PluginName): Observable<Record<PluginName, ServiceStatus>> {
-    const dependencies = this.deps.pluginDependencies.get(plugin);
-    if (!dependencies) {
-      throw new Error(`Unknown plugin: ${plugin}`);
-    }
-    if (!this.dependenciesStatuses.has(plugin)) {
-      this.dependenciesStatuses.set(
-        plugin,
-        this.getPluginStatuses$(dependencies).pipe(
-          // Prevent many emissions at once from dependency status resolution from making this too noisy
-          debounceTime(25)
-        )
-      );
-    }
-    return this.dependenciesStatuses.get(plugin)!;
+    const directDependencies = this.pluginData[plugin].dependencies;
+
+    return this.pluginStatus$.asObservable().pipe(
+      map((allStatus) =>
+        Object.keys(allStatus)
+          .filter((dep) => directDependencies.includes(dep))
```

Comment:
optional nit: the current implementation in the functional style crates a few intermediary arrays. Since this code is hot, I'd consider re-writing it in the imperative style:
```js
      map((allStatus) => {
        const result = {} as Record<PluginName, ServiceStatus>;
        for (const name of Object.keys(allStatus)) {
          if (directDependencies.includes(name)) {
            result[name] = allStatus[name];
          }
        }
        return result;
      }),
```

Reply:
Good point! I'll rewrite it